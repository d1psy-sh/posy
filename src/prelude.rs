pub use std::collections::{HashMap, HashSet};
pub use std::convert::{TryFrom, TryInto};
pub use std::fmt::Display;
pub use std::rc::Rc;
pub use std::str::FromStr;

pub use anyhow::{anyhow, bail, Context, Result};
pub use derivative::Derivative;
pub use once_cell::sync::Lazy;
pub use regex::Regex;
pub use serde::{Deserialize, Serialize};
pub use serde_with::{DeserializeFromStr, SerializeDisplay};
pub use url::Url;
pub use log::{info, trace, warn};

pub use crate::try_from_str_boilerplate;
pub use crate::vocab::*;

use directories::ProjectDirs;
pub static PROJECT_DIRS: Lazy<ProjectDirs> = Lazy::new(|| {
    // ...Can this actually return None?
    ProjectDirs::from("", "Trio Collective", env!("CARGO_PKG_NAME")).unwrap()
});
