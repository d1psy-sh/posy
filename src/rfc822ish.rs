use anyhow::{Context, Result};
use std::collections::HashMap;

// A parsed version of a package METADATA or PKG-INFO or WHEEL file, as per
// https://packaging.python.org/specifications/core-metadata/
pub type Fields = HashMap<String, Vec<String>>;

#[derive(Debug)]
pub struct RFC822ish {
    pub fields: Fields,
    pub body: Option<String>,
}

mod parser_internals {
    use nom::bytes::complete::{is_a, is_not, take_while1};
    use nom::character::complete::one_of;
    use nom::combinator::rest;
    use nom::multi::many1;
    use nom::sequence::separated_pair;
    use nom::{IResult, Parser};
    use nom_supreme::{
        error::ErrorTree,
        final_parser::{final_parser, Location},
        parser_ext::ParserExt,
        tag::complete::tag,
    };

    type ParseResult<'a, O> = IResult<&'a str, O, ErrorTree<&'a str>>;

    // Allegedly, a METADATA file is formatted as an RFC822 email message.
    // This is absolutely not true. The actual format is "whatever
    // the Python stdlib module email.message_from_string does".
    //
    // Overall structure: A series of header lines, then an empty line, then
    // the "message body" (= package readme)
    //
    // A line ending is: \n, \r, or \r\n
    //
    // A header line is:
    // - field name + separator + field value + line ending
    // field names are:
    // - a sequence of one or more characters in the set [\041-\071\073-\176]
    //   or put another way: anything from \041-\176 except for ':'
    // The field name/value separator is:
    // - a colon + zero or more spaces or tabs
    // The field value is:
    // - everything after the separator, until the end of the line, not
    //   including the end of line.
    //   BUT we keep reading if the line ending is followed by a space or tab!
    //   So e.g.:
    //
    //     "foo: \tbar  \n  baz\r\n"
    //
    //   ...will produce the field value:
    //
    //     "bar  \n  baz"
    //
    // Some notable differences from RFC 822:
    // - continuation lines preserve newlines; RFC822 says that they should be
    //   replaced by spaces.
    // - RFC822 says that \r\n is mandatory at the end of all lines
    // - RFC822's definitions of whitespace are a bit different
    //
    // The 'email' module is also extremely lenient of errors. We'll try to be
    // a bit more strict -- we try to be lenient of mangled utf-8, because
    // obviously someone must have messed that up in the history of PyPI, but
    // we fail on oddities like an empty field name or a continuation line at
    // the start of input.

    pub fn parse_metadata(
        input: &str,
    ) -> Result<super::RFC822ish, ErrorTree<Location>> {
        // This has to be an actual function, not just a combinator object,
        // because nom's type system is awkward and if it were a combinator
        // there would be no way to use it multiple times below, even as
        // borrows.
        fn line_ending(input: &str) -> ParseResult<()> {
            tag("\r\n")
                .or(tag("\r"))
                .or(tag("\n"))
                .map(|_| ())
                .parse(input)
        }

        fn is_field_name_char(c: char) -> bool {
            let i = c as u32;
            0o41 <= i && i <= 0o176 && c != ':'
        }

        let field_name = take_while1(is_field_name_char).context("field name");
        let field_separator = tag(":").and(is_a(" \t")).context("field separator");

        let value_line_piece = is_not("\r\n");
        let continuation_marker = line_ending.and(one_of(" \t"));
        let field_value =
            nom::multi::separated_list1(continuation_marker, value_line_piece)
                .recognize()
                .context("field value");

        let field =
            separated_pair(field_name, field_separator, field_value).context("field");
        let fields = many1(field.terminated(line_ending)).context("fields");

        let body = rest.preceded_by(line_ending);

        let metadata = nom::sequence::pair(fields, body.opt());
        let mut parse = final_parser(metadata);
        let (fields_vec, body) = parse(input)?;
        let mut fields = super::Fields::new();
        for (field_name, field_value) in fields_vec {
            fields
                .entry(field_name.to_owned())
                .or_insert(Vec::new())
                .push(field_value.to_owned());
        }
        // Convert from Option<&str> to Option<String>
        let body = body.map(String::from);
        Ok(super::RFC822ish { fields, body })
    }
}

impl RFC822ish {
    pub fn parse(data: &str) -> Result<RFC822ish> {
        parser_internals::parse_metadata(data).context("Error parsing metadata")
    }
}

pub struct CoreMetadata(Fields);

impl CoreMetadata {
    pub fn parse(data: &str) -> Result<CoreMetadata> {
        let mut rfc822ish = RFC822ish::parse(data)?;
        if let Some(body) = rfc822ish.body {
            rfc822ish
                .fields
                .entry("Description".to_string())
                .or_insert(Vec::new())
                .push(body);
        }
        Ok(CoreMetadata(rfc822ish.fields))
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use indoc::indoc;

    #[test]
    fn test_successful_parsing() {
        struct T {
            given: &'static str,
            expected_fields: &'static str,
            expected_body: Option<&'static str>,
        }

        let test_cases = vec![
            T {
                given: indoc! {r#"
                   A: b
                   C: d
                      continued

                   this is the
                   body!
                "#},
                expected_fields: indoc! {r#"
                   {"A": ["b"], "C": ["d\n   continued"]}
                "#},
                expected_body: Some("this is the\nbody!\n"),
            },
            T {
                given: indoc! {r#"
                   no: body
                "#},
                expected_fields: indoc! {r#"
                   {"no": ["body"]}
                "#},
                expected_body: None,
            },
            T {
                given: indoc! {r#"
                   duplicate: one
                   duplicate: two
                   another: field
                   duplicate: three
                "#},
                expected_fields: indoc! {r#"
                   {"duplicate": ["one", "two", "three"], "another": ["field"]}
                "#},
                expected_body: None,
            },
        ];

        for test_case in test_cases {
            let got = RFC822ish::parse(test_case.given).unwrap();
            let expected_fields: Fields =
                serde_json::from_str(test_case.expected_fields).unwrap();
            assert_eq!(got.fields, expected_fields);
            assert_eq!(got.body, test_case.expected_body.map(String::from));
        }
    }

    #[test]
    fn test_failed_parsing() {
        let test_cases = vec![
            "",
            indoc! {r#"
                  continuation line
               at: beginning

               not good
            "#},
            indoc! {r#"
               bad key name: whee
            "#},
            ": no key name\n"
        ];
        for test_case in test_cases {
            let got = RFC822ish::parse(test_case);
            println!("{:?} -> {:?}", test_case, got);
            assert!(got.is_err());
        }
    }
}
