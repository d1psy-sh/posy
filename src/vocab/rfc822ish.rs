use crate::prelude::*;

pub type Fields = HashMap<String, Vec<String>>;

#[cfg_attr(test, derive(Debug, Deserialize, PartialEq))]
pub struct RFC822ish {
    pub fields: Fields,
    pub body: Option<String>,
}

// Allegedly, a METADATA file is formatted as an RFC822 email message.
// This is absolutely not true. The actual format is "whatever
// the Python stdlib module email.parser does". To probe its behavior, a
// convenient entry point is 'email.message_from_string'.
//
// Overall structure: A series of header lines, then an empty line, then
// the "message body" (= description field, in modern PKG-INFO/METADATA
// files).
//
// email.parser module is also extremely lenient of errors. We'll try to be a
// bit more strict -- we try to be lenient of mangled utf-8, because obviously
// someone must have messed that up in the history of PyPI, and aren't picky
// about stuff like trailing newlines. But we fail on oddities like an empty
// field name or a continuation line at the start of input, where email.parser
// would keep on trucking. Fingers crossed that it works out.
peg::parser! {
    grammar rfc822ish_parser() for str {
        // In real RFC822, only CRLF is legal. email.parser is more lenient.
        rule line_ending()
            = quiet!{"\r\n" / "\r" / "\n"}
              / expected!("end of line")

        rule field_name() -> &'input str
            = quiet!{$(['\x21'..='\x39' | '\x3b'..='\x7e']+)}
              / expected!("field name")

        // email.parser drops any " \t" after the colon, but preserves other
        // whitespace in the field value.
        rule field_separator()
            = ":" [' ' | '\t']*

        rule field_value_piece()
            = [^ '\r' | '\n']*

        rule continuation_line_ending()
            = quiet!{line_ending() [' ' | '\t']} / expected!("continuation line")

        // In real RFC822, continuation lines are folded together into a
        // single line, removing the newline characters. email.parser doesn't
        // do that though -- continuation lines just get embedded newlines.
        // (But you don't include any *trailing* newlines. Those are
        // discarded.)
        rule field_value() -> &'input str
            = $(field_value_piece() ** continuation_line_ending())

        rule field() -> (String, String)
            = n:field_name() field_separator() v:field_value()
                { (n.to_owned(), v.to_owned()) }

        rule fields() -> Vec<(String, String)>
            = field() ** line_ending()

        // I think in real RFC822, the body is mandatory? But in early
        // versions of the metadata spec, PKG-INFO/METADATA files didn't have
        // a body, and email.parser don't care, it does what it wants.
        rule trailing_body() -> String
            = line_ending() line_ending() b:$([_]*) { b.to_owned() }

        // The extra line_ending() is to handle the case where there's
        // no trailing body, and exactly one line ending at EOF. If
        // trailing_body matches then the input will be fully consumed by
        // then; if not, then we might have a stray trailing newline to
        // absorb.
        pub rule rfc822ish() -> RFC822ish
            = f:fields() body:(trailing_body()?) line_ending()?
                 {
                     let mut fields = Fields::new();
                     for (name, value) in f {
                         fields.entry(name).or_insert(Vec::new()).push(value)
                     };
                     RFC822ish { fields, body, }
                 }
    }
}

impl RFC822ish {
    pub fn parse(input: &str) -> Result<RFC822ish> {
        Ok(rfc822ish_parser::rfc822ish(input)?)
    }

    pub fn take_all(&mut self, key: &str) -> Vec<String> {
        match self.fields.remove(key) {
            Some(vec) => vec,
            None => Vec::new(),
        }
    }

    pub fn maybe_take_the(&mut self, key: &str) -> Result<Option<String>> {
        let mut values = self.take_all(key);
        match values.len() {
            0 => Ok(None),
            1 => Ok(values.pop()),
            _ => anyhow::bail!("multiple values for singleton key {}", key),
        }
    }

    pub fn take_the(&mut self, key: &str) -> Result<String> {
        match self.maybe_take_the(key)? {
            Some(result) => Ok(result),
            None => anyhow::bail!("can't find required key {}", key),
        }
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use indoc::indoc;

    #[test]
    fn test_successful_parsing() {
        struct T {
            // Input to parser
            given: &'static str,
            // Expected parsed data structure, written as json
            expected: &'static str,
        }

        let test_cases = vec![
            T {
                given: indoc! {r#"
                   A: b
                   C: d
                      continued

                   this is the
                   body!
                "#},
                expected: indoc! {r#"
                   {
                     "fields": { "A": ["b"], "C": ["d\n   continued"]},
                     "body": "this is the\nbody!\n"
                   }
                "#},
            },
            T {
                given: indoc! {r#"
                   no: body
                "#},
                expected: indoc! {r#"
                   {"fields": {"no": ["body"]}}
                "#},
            },
            T {
                given: indoc! {r#"
                   duplicate: one
                   duplicate: two
                   another: field
                   duplicate: three
                "#},
                expected: indoc! {r#"
                   {"fields": {"duplicate": ["one", "two", "three"], "another": ["field"]}}
                "#},
            },
            T {
                given: indoc! {r#"
                no: trailing newline"#},
                expected: indoc! {r#"
                   {"fields": {"no": ["trailing newline"]}}
                "#},
            },
            T {
                given: "",
                expected: r#"{"fields": {}}"#,
            },
        ];

        for test_case in test_cases {
            let got = RFC822ish::parse(test_case.given);
            println!("{:?} -> {:?}", test_case.given, got);
            let got = got.unwrap();
            let expected: RFC822ish = serde_json::from_str(test_case.expected).unwrap();
            assert_eq!(got, expected);
        }
    }

    #[test]
    fn test_failed_parsing() {
        let test_cases = vec![
            indoc! {r#"
                  continuation line
               at: beginning

               not good
            "#},
            indoc! {r#"
               bad key name: whee
            "#},
            ": no key name\n",
        ];
        for test_case in test_cases {
            let got = RFC822ish::parse(test_case);
            println!("{:?} -> {:?}", test_case, got);
            assert!(got.is_err());
        }
    }
}
