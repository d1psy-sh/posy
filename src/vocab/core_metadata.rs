use std::path::Path;

use crate::prelude::*;

use super::rfc822ish::RFC822ish;

/// There are more fields we could add here, but this should be good enough to
/// get started.
#[derive(Debug, Clone)]
#[cfg_attr(test, derive(Serialize))]
pub struct WheelCoreMetadata {
    pub name: PackageName,
    pub version: Version,
    pub requires_dist: Vec<PackageRequirement>,
    pub requires_python: Specifiers,
    pub extras: HashSet<Extra>,
}

#[derive(Debug, Clone)]
#[cfg_attr(test, derive(Serialize))]
pub struct PybiCoreMetadata {
    pub name: PackageName,
    pub version: Version,
    pub environment_marker_variables: HashMap<String, String>,
    pub tags: Vec<String>,
    pub paths: HashMap<String, String>,
}

impl PybiCoreMetadata {
    pub fn path(&self, key: &str) -> Result<&Path> {
        Ok(self.paths.get(key).ok_or(anyhow!("bad pybi: no '{key}' path"))?.as_ref())
    }
}

fn parse_common(input: &[u8]) -> Result<(PackageName, Version, RFC822ish)> {
    let input = String::from_utf8_lossy(input);
    let mut parsed = RFC822ish::parse(&input)?;

    static NEXT_MAJOR_METADATA_VERSION: Lazy<Version> =
        Lazy::new(|| "3".try_into().unwrap());

    // Quoth https://packaging.python.org/specifications/core-metadata:
    // "Automated tools consuming metadata SHOULD warn if metadata_version
    // is greater than the highest version they support, and MUST fail if
    // metadata_version has a greater major version than the highest
    // version they support (as described in PEP 440, the major version is
    // the value before the first dot)."
    //
    // We do the MUST, but I think I disagree about warning on
    // unrecognized minor revisions. If it's a minor revision, then by
    // definition old software is supposed to be able to handle it "well
    // enough". The only purpose of the warning would be to alert users
    // that they might want to upgrade, or to alert the tool authors that
    // there's a new metadata release. But for users, there are better
    // ways to nudge them to upgrade (e.g. checking on startup, like
    // pip does), and new metadata releases are so rare and so
    // much-discussed beforehand that if a tool's authors don't know
    // about it it's because the tool is abandoned anyway.
    let metadata_version: Version = parsed.take_the("Metadata-Version")?.try_into()?;
    if metadata_version >= *NEXT_MAJOR_METADATA_VERSION {
        bail!("unsupported Metadata-Version {}", metadata_version);
    }

    Ok((
        parsed.take_the("Name")?.parse()?,
        parsed.take_the("Version")?.try_into()?,
        parsed,
    ))
}

impl TryFrom<&[u8]> for WheelCoreMetadata {
    type Error = anyhow::Error;

    fn try_from(value: &[u8]) -> Result<Self, Self::Error> {
        let (name, version, mut parsed) = parse_common(value)?;

        let mut requires_dist = Vec::new();
        for req_str in parsed.take_all("Requires-Dist").drain(..) {
            requires_dist.push(req_str.try_into()?);
        }

        let requires_python = match parsed.maybe_take_the("Requires-Python")? {
            Some(rp_str) => rp_str.try_into()?,
            None => Specifiers(Vec::new()),
        };

        let mut extras: HashSet<Extra> = HashSet::new();
        for extra in parsed.take_all("Provides-Extra").drain(..) {
            extras.insert(extra.parse()?);
        }

        Ok(WheelCoreMetadata {
            name,
            version,
            requires_dist,
            requires_python,
            extras,
        })
    }
}

impl TryFrom<&[u8]> for PybiCoreMetadata {
    type Error = anyhow::Error;

    fn try_from(value: &[u8]) -> Result<Self, Self::Error> {
        let (name, version, mut parsed) = parse_common(value)?;

        Ok(PybiCoreMetadata {
            name,
            version,
            environment_marker_variables: serde_json::from_str(
                &parsed.take_the("Pybi-Environment-Marker-Variables")?,
            )?,
            tags: parsed.take_all("Pybi-Wheel-Tag"),
            paths: serde_json::from_str(&parsed.take_the("Pybi-Paths")?)?,
        })
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use indoc::indoc;

    #[test]
    fn test_basic_core_parse() {
        let metadata_text = indoc! {r#"
            Metadata-Version: 2.1
            Name: trio
            Version: 0.16.0
            Summary: A friendly Python library for async concurrency and I/O
            Classifier: Framework :: Trio
            Requires-Python: >=3.6
            Requires-Dist: attrs (>=19.2.0)
            Requires-Dist: sortedcontainers
            Requires-Dist: contextvars[foo] (>=2.1) ; python_version < "3.7"

            The Trio project's goal is...
        "#}
        .as_bytes();

        let metadata: WheelCoreMetadata = metadata_text.try_into().unwrap();

        insta::assert_ron_snapshot!(metadata, @r###"
        WheelCoreMetadata(
          name: "trio",
          version: "0.16.0",
          requires_dist: [
            "attrs >= 19.2.0",
            "sortedcontainers",
            "contextvars[foo] >= 2.1; python_version < \"3.7\"",
          ],
          requires_python: ">= 3.6",
          extras: [],
        )
        "###);
    }

    #[test]
    fn test_basic_pybi_parse() {
        let metadata_text = indoc! {r#"
            Metadata-Version: 2.1
            Name: CPython
            Version: 3.11.2
            Pybi-Environment-Marker-Variables: {"implementation_name": "cpython", "os_name": "posix"}
            pybi-wheel-tag: cp311-cp311-linux_x86_64
            Pybi-Wheel-Tag: py3-none-any
            Pybi-Paths: {"data": ".", "include": "include/python3.11"}

            This is CPython, the standard interpreter for the Python language...
        "#}
        .as_bytes();

        let metadata: PybiCoreMetadata = metadata_text.try_into().unwrap();

        insta::assert_ron_snapshot!(metadata,
            {
                ".paths" => insta::sorted_redaction(),
                ".environment_marker_variables" => insta::sorted_redaction(),
            },
                                    @r###"
        PybiCoreMetadata(
          name: "CPython",
          version: "3.11.2",
          environment_marker_variables: {
            "implementation_name": "cpython",
            "os_name": "posix",
          },
          tags: [
            "cp311-cp311-linux_x86_64",
            "py3-none-any",
          ],
          paths: {
            "data": ".",
            "include": "include/python3.11",
          },
        )
        "###
        );
    }
}
