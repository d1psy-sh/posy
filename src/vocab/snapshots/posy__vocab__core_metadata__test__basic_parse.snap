---
source: src/vocab/core_metadata.rs
expression: metadata

---
CoreMetadata {
    metadata_version: Version(
        Version {
            epoch: 0,
            release: [
                2,
                1,
            ],
            pre: None,
            post: None,
            dev: None,
            local: [],
        },
    ),
    name: PackageName {
        as_given: "trio",
        normalized: "trio",
    },
    version: Version(
        Version {
            epoch: 0,
            release: [
                0,
                16,
                0,
            ],
            pre: None,
            post: None,
            dev: None,
            local: [],
        },
    ),
    requires_dist: [
        PackageRequirement(
            Requirement {
                name: PackageName {
                    as_given: "attrs",
                    normalized: "attrs",
                },
                extras: [],
                specifiers: Specifiers(
                    [
                        Specifier {
                            op: GreaterThanEqual,
                            value: "19.2.0",
                        },
                    ],
                ),
                env_marker: None,
            },
        ),
        PackageRequirement(
            Requirement {
                name: PackageName {
                    as_given: "sortedcontainers",
                    normalized: "sortedcontainers",
                },
                extras: [],
                specifiers: Specifiers(
                    [],
                ),
                env_marker: None,
            },
        ),
        PackageRequirement(
            Requirement {
                name: PackageName {
                    as_given: "contextvars",
                    normalized: "contextvars",
                },
                extras: [
                    Extra(
                        PackageName {
                            as_given: "foo",
                            normalized: "foo",
                        },
                    ),
                ],
                specifiers: Specifiers(
                    [
                        Specifier {
                            op: GreaterThanEqual,
                            value: "2.1",
                        },
                    ],
                ),
                env_marker: Some(
                    Operator {
                        op: Compare(
                            StrictlyLessThan,
                        ),
                        lhs: Variable(
                            "python_version",
                        ),
                        rhs: Literal(
                            "3.7",
                        ),
                    },
                ),
            },
        ),
    ],
    requires_python: Specifiers(
        [
            Specifier {
                op: GreaterThanEqual,
                value: "3.6",
            },
        ],
    ),
    extras: {},
}
